tests/beta/tapes/test_tape.py

```diff
@@ -994,3 +994,533 @@ def test_multiple_output_values(self, tol):
 
         res = tape.execute(dev)
         assert res.shape == (2,)
+
+
+class TestGradMethod:
+    """Tests for parameter gradient methods"""
+
+    def test_non_differentiable(self):
+        """Test that a non-differentiable parameter is
+        correctly marked"""
+        psi = np.array([1, 0, 1, 0]) / np.sqrt(2)
+
+        with QuantumTape() as tape:
+            qml.QubitStateVector(psi, wires=[0, 1])
+            qml.RX(0.543, wires=[0])
+            qml.RY(-0.654, wires=[1])
+            qml.CNOT(wires=[0, 1])
+            probs(wires=[0, 1])
+
+        assert tape._grad_method(0) is None
+        assert tape._grad_method(1) == "F"
+        assert tape._grad_method(2) == "F"
+
+        tape._update_gradient_info()
+
+        assert tape._par_info[0]["grad_method"] is None
+        assert tape._par_info[1]["grad_method"] == "F"
+        assert tape._par_info[2]["grad_method"] == "F"
+
+    def test_independent(self):
+        """Test that an independent variable is properly marked
+        as having a zero gradient"""
+
+        with QuantumTape() as tape:
+            qml.RX(0.543, wires=[0])
+            qml.RY(-0.654, wires=[1])
+            expval(qml.PauliY(0))
+
+        assert tape._grad_method(0) == "F"
+        assert tape._grad_method(1) == "0"
+
+        tape._update_gradient_info()
+
+        assert tape._par_info[0]["grad_method"] == "F"
+        assert tape._par_info[1]["grad_method"] == "0"
+
+        # in non-graph mode, it is impossible to determine
+        # if a parameter is independent or not
+        tape._graph = None
+        assert tape._grad_method(1, use_graph=False) == "F"
+
+
+class TestJacobian:
+    """Unit tests for the jacobian method"""
+
+    def test_unknown_grad_method_error(self):
+        """Test error raised if gradient method is unknown"""
+        tape = QuantumTape()
+        with pytest.raises(ValueError, match="Unknown gradient method"):
+            tape.jacobian(None, method="unknown method")
+
+    def test_non_differentiable_error(self):
+        """Test error raised if attempting to differentiate with
+        respect to a non-differentiable argument"""
+        psi = np.array([1, 0, 1, 0]) / np.sqrt(2)
+
+        with QuantumTape() as tape:
+            qml.QubitStateVector(psi, wires=[0, 1])
+            qml.RX(0.543, wires=[0])
+            qml.RY(-0.654, wires=[1])
+            qml.CNOT(wires=[0, 1])
+            probs(wires=[0, 1])
+
+        # by default all parameters are assumed to be trainable
+        with pytest.raises(
+            ValueError, match=r"Cannot differentiate with respect to parameter\(s\) {0}"
+        ):
+            tape.jacobian(None)
```

Comment:
Wouldn't that just set the device to ``None``, which should raise a different error?

Reply:
Oh, the error resulting from the device being 'none' occurs later than the error I'm catching above 😆 Hence why it doesn't show up here. I think I was being a bit lazy